gnome3.gnome-settings-daemon: fix #14168

worldofpeace opened the pull request:
###### Motivation for this change
Originally from https://github.com/NixOS/nixpkgs/pull/54584 then got moved to the 3.32 pr.

However GitHub hides long discussions in prs that get really long, like the one's for major GNOME3 updates. So putting bugfixes in there, where discussion is important to be easily accessible, isn't a good idea.

~~This is blocked by https://github.com/NixOS/nixpkgs/pull/57027 because it depends on code added
in that release.~~

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

hedning commented on nixos/modules/services/desktops/gnome3/gnome-settings-daemon.nix:
```diff
@@ -40,6 +40,8 @@ in
 
     services.udev.packages = [ cfg.package ];
 
+    security.wrappers.gsd-backlight-helper.source = "${pkgs.gnome3.gnome-settings-daemon}/libexec/gsd-backlight-helper";
```
I'm getting this when running the wrapper:

>❯ gsd-backlight-helper 
This program can only be used by the root user

So we're hitting this for some reason: https://gitlab.gnome.org/GNOME/gnome-settings-daemon/blob/master/plugins/power/gsd-backlight-helper.c#L65

Infinisil replied:
Is this resolved with the latest version of this PR?

worldofpeace replied:
I've left this alone for a bit because we just need to get 3.32 on master first.
Then debugging would be quick.

Though I'm not sure why we're getting https://gitlab.gnome.org/GNOME/gnome-settings-daemon/blob/master/plugins/power/gsd-backlight-helper.c#L65

worldofpeace commented:
Huh, so no matter what I've tried to do there's still https://gitlab.gnome.org/GNOME/gnome-settings-daemon/blob/master/plugins/power/gsd-backlight-helper.c#L65 where `uid` and `euid` are not `0`.

I've checked and `getuid ()` for sure returns the `uid` of the non-privileged user and not root.
Don't know why setuid didn't do the thing.

Perhaps we can do a worse hack where we symlink `gsd-backlight-helper` to `bin` and change the path in the polkit policy to `/run/current-system/sw/bin/gsd-backlight-helper`?

worldofpeace commented:
So I've looked and I don't know why this isn't working.
I locally added `setuid = true;` and `user`, `group` to `root`, and with tests `getuid ()` still returned `1000` for me.

Think something like https://github.com/worldofpeace/nixpkgs/commit/c744359579d4d3e2e705d75b4383673b9347a984 would work?

hedning commented:
Yeah, seems like `getuid()` simply returns the actual user id (as oppsosed to `geteuid()` which returns the _effective_ id) so we'd have to patch gsd for it to be used with setuid.

Seeing that the problem is a new polkit policy file, but an old gsd application (https://github.com/NixOS/nixpkgs/issues/14168#issuecomment-381759563) your fix should hopefully work. Seems sane to me :+1:

worldofpeace commented:
> Yeah, seems like getuid() simply returns the actual user id (as oppsosed to geteuid() which returns the effective id) so we'd have to patch gsd for it to be used with setuid.

Strangely enough the effective id would only be 0 if I removed the layer of wrapping from `wrapGAppsHook`
(i think).

> Seeing that the problem is a new polkit policy file, but an old gsd application (#14168 (comment)) your fix should hopefully work. Seems sane to me +1

Right, think I'll go with that since we'll still be able to use polkit.

worldofpeace committed:
gnome3.gnome-settings-daemon: fix #14168

The issue in brief is caused by the fact that gnome-settings-daemon is
non-restartable and if the policy definitions change in any way
it won't propagate to the running system because of that.

By patching references to the backlight helper to
/run/current-system/sw/bin/gsd-backlight-helper
the polkit policy will not undergo a change.

hedning reviewed (approved):
Haven't tested, but LGTM

jtojnar commented on pkgs/desktops/gnome-3/core/gnome-settings-daemon/default.nix:
```diff
@@ -93,6 +94,12 @@ stdenv.mkDerivation rec {
+  # So the polkit policy can reference /run/current-system/sw/bin/gsd-backlight-helper
+  postFixup = ''
+    mkdir $out/bin
+    ln -s $out/libexec/gsd-backlight-helper $out/bin/gsd-backlight-helper
```
Can we use some other path that is guaranteed to be linked but not in the path?

jtojnar replied:
Maybe a subdirectory of `/bin`

worldofpeace replied:
Doing `bin/gnome-settings-daemon`.

hedning replied:
Could `pathsToLink = [ "/libexec/gsd-backlight-helper" ];` `gnome-settings-daemon.nix` work? Seems somewhat cleaner to me, though not a big deal.

worldofpeace replied:
Yeah thought of that too. It'd be cleaner but I don't see it happening soon where `gnome-settings-daemon` wouldn't be in `environment.systemPackages`. (thusly needing that).
